lib/Conversion/HWToSystemC/HWToSystemC.cpp

```diff
@@ -21,6 +22,112 @@
+/// the body of the module into the new SystemC module by splitting up the body
+/// into field declarations, initializations done in a newly added systemc.ctor,
+/// and internal methods to be registered in the constructor.
+struct ConvertHWModule : public OpConversionPattern<HWModuleOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(HWModuleOp module, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    // Collect the HW module's port types.
+    FunctionType moduleType = module.getFunctionType();
+    TypeRange moduleInputs = moduleType.getInputs();
+    TypeRange moduleOutputs = moduleType.getResults();
+
+    // SystemC module port types are all expressed as block arguments to the op,
+    // so collect all of the types and add them as arguments to the SystemC
+    // module.
+    SmallVector<Type, 4> portTypes(moduleInputs);
+    portTypes.append(moduleOutputs.begin(), moduleOutputs.end());
+
+    // Collect all the port directions.
+    SmallVector<systemc::PortDirection> directions;
+    for (auto port : module.getAllPorts()) {
+      if (port.isInput())
+        directions.push_back(systemc::PortDirection::Input);
+      else if (port.isOutput())
+        directions.push_back(systemc::PortDirection::Output);
+      else
+        directions.push_back(systemc::PortDirection::InOut);
+    }
+    PortDirectionsAttr portDirections =
+        PortDirectionsAttr::get(rewriter.getContext(), directions);
+
+    // Collect all the port names (inputs and outputs).
+    SmallVector<Attribute> portNames;
+    ArrayRef<Attribute> args = module.getArgNames().getValue();
+    ArrayRef<Attribute> results = module.getResultNames().getValue();
+    portNames.append(args.begin(), args.end());
+    portNames.append(results.begin(), results.end());
+
+    // Create the SystemC module. Note that no parameterized modules are
+    // supported yet.
+    auto scModule = rewriter.create<SCModuleOp>(
+        module.getLoc(), portDirections,
+        ArrayAttr::get(rewriter.getContext(), portNames));
+
+    // Inline the HW module body into the SystemC module body.
+    Region &scModuleBodyRegion = scModule.getBodyRegion();
+    rewriter.inlineRegionBefore(module.getBodyRegion(), scModuleBodyRegion,
+                                scModuleBodyRegion.end());
+
+    // Set the SCModule type and name attributes. Add block arguments for each
+    // output.
+    auto scModuleType = rewriter.getFunctionType(portTypes, {});
+    rewriter.updateRootInPlace(scModule, [&] {
+      scModule->setAttr(scModule.getTypeAttrName(),
+                        TypeAttr::get(scModuleType));
+      scModule.setName(module.getName());
+      scModuleBodyRegion.addArguments(
+          moduleOutputs, SmallVector<Location, 4>(moduleOutputs.size(),
+                                                  rewriter.getUnknownLoc()));
+    });
+
+    // Erase the HW module.
+    rewriter.eraseOp(module);
```

Comment:
Is it safe to erase the HW module at this point given that it may have uses (instantiations) in another HW module?

Reply:
If you lower one isolated design in its entirety, it is save (will error out on the instance not being lowerable). When implementing the instance lowering we have to make sure that it is aware of the referred module op already being lowered. If we only lower one (or a subset) of module to systemc, or there is an instantiation in another design refering to a module in this one via an external module, then this implementation is not save. But partially lowering a design is not supported yet and for the other case we need to implement some interop, which I would leave for a later PR. So I would suggest we erase if for now and don't consider such external references for now so we don't have to deal with name uniquification, etc. in this PR.